File: tree/865所有具有最深节点的最小子树.py
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def subtreeWithAllDeepest(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        max_depth = 0
        res = None
        def dfs1(root,depth) -> int:
            nonlocal max_depth,res
            if not root:
                max_depth = max(max_depth,depth)
                return depth
            left = dfs1(root.left,depth+1)
            right = dfs1(root.right,depth+1)
            if left == max_depth and right == max_depth:
                res = root
            return max(left,right)
        dfs1(root,0)
        return res

        

File: tree/865所有具有最深节点的最小子树.cpp
#include "binarytree.h"
#include <iostream>
using namespace std;
class Solution {
public:
    TreeNode*ans;
    int maxdepth=0;
    TreeNode* subtreeWithAllDeepest(TreeNode* root) {
        dfs(root,0);
        return ans;
    }
    int dfs(TreeNode* root,int depth){
        if(root==nullptr){
            maxdepth=max(maxdepth,depth);
            return depth;
        }
        
        int left=dfs(root->left,depth+1);
        int right=dfs(root->right,depth+1);
        if(left==maxdepth&&right==maxdepth){
            ans=root;
        }
        return max(left,right);
    }
};